core/src/sumeragi/mod.rs

```diff
@@ -127,22 +143,76 @@ impl Sumeragi {
             .internal
             .wsv
             .lock()
-            .expect("Failed to lock on `update_metrics`. Mutex poisoned");
+            .expect("Failed to lock on `wsv`. Mutex poisoned");
+
+        let metrics_guard = self
+            .metrics_mutex
+            .lock()
+            .expect("Failed to lock on `metrics`. Mutex poisoned");
+
+        let mut last_block_height_guard = self
+            .last_block_height
+            .lock()
+            .expect("Failed to lock on `last_block_height_when_updating_metrics`. Mutex poisoned");
+
+        let start_index = *last_block_height_guard;
+        {
+            let blocks_iter = wsv_guard.blocks();
+            let blocks_iter =
+                blocks_iter.skip(start_index.try_into().expect("Failed to cast to u32."));
+            for block in blocks_iter {
+                let block_txs_accepted = block.as_v1().transactions.len() as u64;
+                let block_txs_rejected = block.as_v1().rejected_transactions.len() as u64;
+
+                metrics_guard
+                    .txs
+                    .with_label_values(&["accepted"])
+                    .inc_by(block_txs_accepted);
+                metrics_guard
+                    .txs
+                    .with_label_values(&["rejected"])
+                    .inc_by(block_txs_rejected);
+                metrics_guard
+                    .txs
+                    .with_label_values(&["total"])
+                    .inc_by(block_txs_accepted + block_txs_rejected);
+                metrics_guard.block_height.inc();
+            }
+            *last_block_height_guard = wsv_guard.height();
+        }
+
+        metrics_guard.domains.set(wsv_guard.domains().len() as u64);
+
+        let mut last_metric_tx_amounts_guard = self
+            .last_metric_tx_amounts
+            .lock()
+            .expect("Failed to lock on `last_metric_tx_amounts`. Mutex poisoned");
+        let mut last_metric_tx_amounts_counter_guard = self.last_metric_tx_amounts_counter.lock().expect("Failed to lock on `last_total_tx_amounts_counter_when_updating_metrics`. Mutex poisoned");
+
+        let diff_count =
+            wsv_guard.metric_tx_amounts_counter.get() - *last_metric_tx_amounts_counter_guard;
+        let diff_amount_per_count = (wsv_guard.metric_tx_amounts.get()
+            - *last_metric_tx_amounts_guard)
+            / (diff_count as f64);
```

Comment:
Here you're explicitly casting `diff_count`, but `wsv.metric_tx_amounts` is already `f64`, even though here's the only place where it's needed. I'd rather have it `u64` like everything else and cast it here as this field cannot be fractional by itself AFAICT.

Reply:
The cast to f64 is required because otherwise the operation is `f64 / u64` which is not legal in rust.

Reply:
I made that mistake too. Rust is strictly typed.